keps/79-hierarchical-cohorts/README.md

```diff
@@ -0,0 +1,220 @@
+* Allow setting constraints about borrowing and lending at all levels.
+* Provide quota for groups of Queues.
+
+### Non-Goals
+
+* Change the existing API and mechanics in a not backward-compatible way.
+* Introduce a new, separate API.
+* Introduce new ways of fair sharing, like ratio-based sharing (at least 
+not in this KEP).
+* Introduce additional preemption models (this will be in a separate KEP).
+
+## Proposal
+
+Introduce a new object called Cohort with the similar quota provisioning mechanism
+as ClusterQueue. Chorots additionally may specify its parent, another Cohort,
+forming together a tree-like organization structure. ClusterQueues will still be able to specify
+the Cohort they belong to. The Cohort mentioned by ClusterQueue doesn't require
+an actual object to be present. If such is not provided, it is understood 
+that the Cohort doesn't provide any quota, has no parent, doesn't belong to any bigger
+structure or has any non-default settings.
+
+The difference between ClusterQuota and Cohort will be that:
+
+* ClusterQueues are leaves in the organization tree, Cohorts are inner nodes. 
+* Cohort doesn't accept any workloads.
+* Quota provided at the Cohort level is to be shared with the entire organization and doesn't 
+  have an owning ClusterQueue.
+* Borrowing limit specified at the Cohort level means that the entire subtree cannot borrow more 
+  from the rest of the organization tree than the given value.
+* Lending limit specified at the Cohort levels means that the rest of the organization tree
+  cannot borrow more from the subtree than the given value.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I have two multi-team organizations in the company. One that does research and one that runs production
+workloads. Both are given some quota that is further distributed among the subteams. I want to grant 
+the production workloads the ability to borrow research quota if needed, but not the other way round.
+
+#### Story 2
+
+I have a couple organizations that have dedicated resources. The organizations should not borrow 
+from each other, however I want to have an additional "special" queue, with low priority jobs, that can 
+borrow unused capacity from any of the organizations.
+
+### Risks and Mitigations
+
+* Users may not create a Cohort object for Cohorts mentioned in ClusterQueues - this is fine,
+we will assume that the Cohort doesn't have any quota or other constraints/settings. 
+
+* Users may provide non-existent parent information for a Cohort - this is fine, 
+we will simply assume that the parent has no quota or other constraints/settings.
+
+* Users may create a cycle in the Cohort hierarchy - Kueue will stop all new admissions within
+the entire tree. The already admitted workloads will be allowed to continue. Appropriate 
+ClusterQueue/Cohort Status Conditions will be set and Events emited.
+
+* Scheduling and preemption may require more computation/resources.
+
+## Design Details
```

Comment:
If we have `ClusterQueue.preemption.reclaimWithinCohort=Any`, does it mean preemption is allowed within the first level of cohort, or the whole hierarchy?

Reply:
Whole hierarchy.

Reply:
I would suggest adding such a note on preemption in KEP.

Reply:
Added to proposal section.